raster-foundry/raster-foundry
# Implement analyses sharing

## Overview

In order to share an analysis, we needed a way to also share the projects within that analysis.
Instead of creating and managing ACRs, this implementation, when the `analysisId` query parameter is present, evaluates project access authorization with respect to the analysis. In order to get a project or the datasources for that project using an `analysisId`, the following conditions are enforced:

- the user must have access to the analysis the id indicates
- the analysis must contain the project that is being fetched
- the owner of the analysis must be authorized to `VIEW` the project in question 

This PR addresses the various parts needed to implement the above strategy:

**Backend**:
- adds an `analysisId` query parameter to the `GET projects/{projectId}` and `GET projects/{projectId}/datasources` endpoints that opts in to authentication within the provided analyses' context

**Frontend**:
- modifies all requests within the lab context that require auth to utilize the `analysisId` parameter
- updates some `project.service` methods to allow the passing of extra parameters
- adds a sub-nav-bar to the lab to match the project editor and provide a good spot for the permissions option
- adds a prop to the lab-node component to allow the disabling of node sharing when someone other than the owner is viewing the analysis
- updates eslint and our config to allow for spread properties on object literals (`{...objA, ...objB, someProp}`)
- updates eslint config to allow compilation with errors in development while failing builds in non development environments
- fixes a bug with the input node that would not initialize the band values for datasources that don't have bands defined

### Checklist

- [NA] Styleguide updated, if necessary
- [NA] Symlinks from new migrations present or corrected for any new migrations
- [NA] Any content changes are properly templated using `BUILDCONFIG.APP_NAME`
- [NA] Any new SQL strings have tests

### Demo

New lab sub-navbar
![image](https://user-images.githubusercontent.com/2442245/50319757-eb40aa80-0496-11e9-9112-3833392d596b.png)


### Notes

- See paired spec PR https://github.com/raster-foundry/raster-foundry-api-spec/pull/73

- This PR is based on #4357

- Still need to update the spec to reflect the added QP

- More work will need to be done to properly hide all of the various interactions that would not be usable when a user has read-only access

## Testing Instructions

This should be tested with two user accounts (A and B in the instructions). When testing I used the dev user and my own google social login.

 * User A should be able to work with an analysis that they own as usual
 * The 'Permissions' option should open the modal
 * Adding permissions and then closing the modal should fire off the expected requests which should complete successfully
 * Add permission for user B to an analysis that you have confirmed works normally
 * Log into the app with user B and changing the analysis browsing ownership filter should display the analysis that was just shared
 * User B should be able to view the analysis as expected (if only view permission were given, they cannot change bands or modify anything)

To test the security features:
* Make or use a project from User B and copy the id somewhere. This project should **NOT** be shared with User A
* Log in as user A and open an analysis; then modify something about the analysis and copy the PUT request
* Replace a project id within the `executionParameters` with the project id from User B (which the current user does not have access to
* Open the analysis and notice that the request for that project now fails with a 403 and no info about the project is displayed in the UI

Closes #4352
